### PR TITLE
3 module tdd

### PR DESCRIPTION
**Pull request by maxonchil**
* Added main functions into app using TDD;
* Added documentation for main functions;
* Extended app with some new types and interfaces;
* Added interface of BaseNode which include id typeof number, to extend from it;
* Added separate state for challenge;
* Fixed some esling rules;
* Removed prettier from app and added lodash;
* Added mock filed with achievements and tasks;
* Added json module resolution to make possible import json files into my ts files;

### CODE REVIEW

File: src/utils/calculateAchievementsStatus/calculateAchievementsStatus.function.spec.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	mockedAchievementsStatusMap,
+	mockedSuccessTasksStatusMap,
+	successStatus, pendingStatus, mockedFailTasksStatusMap, failStatus,
+} from './calculateAchievementsStatus.mock';
+
+
+
+
+describe('#calculateAchievementsStatus', () => {
+	describe('Arguments', () => {
```
Comment (oleksandr-zhynzher) [inline src/utils/calculateAchievementsStatus/calculateAchievementsStatus.function.spec.ts]: I think it`s better to rename this describe to "With invalid arguments"

File: src/utils/calculateAchievementsStatus/calculateAchievementsStatus.function.spec.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+			it('should return empty object, if tasks statuses is an empty object', () => {
+				// @ts-ignore
+				const createdAchievement = calculateAchievementsStatus(mockedAchievements, {});
+				expect(createdAchievement).toEqual({});
+			});
+		});
+	});
+
+	describe('Main logic', () => {
```
Comment (oleksandr-zhynzher) [inline src/utils/calculateAchievementsStatus/calculateAchievementsStatus.function.spec.ts]: Better to rename this describe to "With valid arguments"

File: src/utils/calculateAchievementsStatus/calculateAchievementsStatus.function.spec.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+		it('should return achievements status map for the challenge', () => {
+			const createdAchievement = calculateAchievementsStatus(mockedAchievements, mockedTasksStatusMap);
```
Comment (oleksandr-zhynzher) [inline src/utils/calculateAchievementsStatus/calculateAchievementsStatus.function.spec.ts:43]: Let`s move this const to beforeEach

File: src/utils/checkComplete/checkComplete.function.spec.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
+import { checkComplete } from './checkComplete.function';
+import { TasksStatusMap } from '../../types';
+import { StatusStates } from '../../enums';
+
+const mockedTasksStatus: TasksStatusMap = {
```
Comment (oleksandr-zhynzher) [inline src/utils/checkComplete/checkComplete.function.spec.ts]: lets move all this const to beforeEach

File: src/utils/checkComplete/checkComplete.function.spec.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+		updated: new Date().toLocaleDateString(),
+	},
+	4: {
+		state: StatusStates.Success,
+		updated: new Date().toLocaleDateString(),
+	},
+};
+
+describe('#checkComplete', () => {
+	describe('Arguments', () => {
```
Comment (oleksandr-zhynzher) [inline src/utils/checkComplete/checkComplete.function.spec.ts]: The same issue with naming

File: src/utils/getTaskForToday/getTaskForToday.function.spec.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+		it('should return task for today if today is a first day of challenge', () => {
+			expect(getTaskForToday(1, mockedChallenges)).toEqual(mockedTaskForToday);
+		});
+
+		it('should return task for today, if today is 2nd day of challenge', () => {
+			expect(getTaskForToday(4, mockedChallenges)).toEqual(mockedTaskForToday);
+		});
+
+		it('should return empty object, if no challenge with such id', () => {
```
Comment (oleksandr-zhynzher) [inline src/utils/getTaskForToday/getTaskForToday.function.spec.ts:69]: But you already test it

File: src/utils/getTaskForToday/getTaskForToday.function.spec.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+	describe('Corner cases', () => {
+		it('should return empty object if challenge is completed', () => {
+			expect(getTaskForToday(2, mockedChallenges)).toEqual({});
+		});
+
+		it('should return empty object if challenge not started', () => {
+			expect(getTaskForToday(3, mockedChallenges)).toEqual({});
+		});
+
+		it('should return empty object if no task for today', () => {
```
Comment (oleksandr-zhynzher) [inline src/utils/getTaskForToday/getTaskForToday.function.spec.ts]: It's looks like the same test case like `should return empty object if no challenge with such id`

File: src/utils/isSomeChallangeInProgress/isSomeChallengeInProgress.function.spec.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+];
+
+describe('#isSomeChallangeInProgress', () => {
+	describe('Main logic', () => {
+		it('should return true if some of passed challenges in progress', () => {
+			expect(isSomeChallengeInProgress(mockedProgressChallenges)).toBeTruthy();
+		});
+
+		it('should return false if no one of passed challenges in progress', () => {
+			expect(isSomeChallengeInProgress(mockedFinishedChallenges)).toBeTruthy();
```
Comment (oleksandr-zhynzher) [inline src/utils/isSomeChallangeInProgress/isSomeChallengeInProgress.function.spec.ts:31]: according to description need to use `.toBeFalsy()`

File: src/utils/startNewChallenge/startNewChallenge.function.spec.ts [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+		actualAchievements: mockedActualAchievements,
+		achievementsStatus: mockedAchievementStatus,
+	} as Challenge,
+];
+
+describe('#startNewChallenge', () => {
+	describe('Arguments', () => {
+		describe('Empty tasks', () => {
+			it('should return empty object, if no tasks list was passed as an argument', () => {
+				const createdChallenge = startNewChallenge([], mockedChallenges, 0, 0);
```
Comment (oleksandr-zhynzher) [inline src/utils/startNewChallenge/startNewChallenge.function.spec.ts:61]: If you are using TS, you can check the type of the arguments by TS mechanism and we don't need to test in our test cases, or I missing something?
